px4_ros2_cpp/src/navigation/experimental/local_navigation_interface.cpp

```diff
@@ -0,0 +1,201 @@
+/****************************************************************************
+ * Copyright (c) 2023 PX4 Development Team.
+ * SPDX-License-Identifier: BSD-3-Clause
+ ****************************************************************************/
+
+#include <rclcpp/rclcpp.hpp>
+#include <px4_ros2/navigation/experimental/local_navigation_interface.hpp>
+
+namespace px4_ros2
+{
+
+LocalNavigationInterface::LocalNavigationInterface(
+  rclcpp::Node & node, const uint8_t pose_frame,
+  const uint8_t velocity_frame)
+: _node(node),
+  _pose_frame(pose_frame),
+  _velocity_frame(velocity_frame)
+{
+  // Check that selected pose and velocity reference frames are valid
+  auto it_pose_frame = std::find(
+    std::begin(_available_pose_frames), std::end(
+      _available_pose_frames), pose_frame);
+  if (it_pose_frame == std::end(_available_pose_frames)) {
+    RCLCPP_ERROR(
+      node.get_logger(), "Failed to initialize LocalNavigationInterface: invalid pose reference frame %i.",
+      pose_frame);
+    rclcpp::shutdown();
+  }
+
+  auto it_vel_frame =
+    std::find(
+    std::begin(_available_velocity_frames), std::end(
+      _available_velocity_frames), velocity_frame);
+  if (it_vel_frame == std::end(_available_velocity_frames)) {
+    RCLCPP_ERROR(
+      node.get_logger(), "Failed to initialize LocalNavigationInterface: invalid velocity reference frame %i.",
+      velocity_frame);
+    rclcpp::shutdown();
+  }
+
+  _aux_local_position_pub = node.create_publisher<AuxLocalPosition>(AUX_LOCAL_POSITION_TOPIC, 10);
+
+}
+
+int LocalNavigationInterface::update(const LocalPositionEstimate & local_position_estimate) const
+{
+  // Run basic sanity checks on local position estimate
+  if (_checkEstimateEmpty(local_position_estimate)) {
+    RCLCPP_WARN(_node.get_logger(), "Estimate values are all empty.");
+    return static_cast<int>(NavigationInterfaceCodes::ESTIMATE_EMPTY);
+  }
+
+  if (!_checkVarianceValid(local_position_estimate)) {
+    return static_cast<int>(NavigationInterfaceCodes::ESTIMATE_VARIANCE_INVALID);
+  }
+
+  if (!_checkFrameValid(local_position_estimate)) {
+    return static_cast<int>(NavigationInterfaceCodes::ESTIMATE_FRAME_UNKNOWN);
+  }
+
+  if (local_position_estimate.timestamp_sample == 0) {
+    RCLCPP_WARN(_node.get_logger(), "Estimate sample timestamp is empty.");
+  }
+
+  // Populate aux local position message
+  AuxLocalPosition aux_local_position;
+
+  aux_local_position.timestamp_sample = local_position_estimate.timestamp_sample;
+
+  // Position
+  aux_local_position.pose_frame = _pose_frame;
+
+  const Vector2f position_xy = local_position_estimate.position_xy.value_or(Vector2f{NAN, NAN});
+  aux_local_position.position[0] = position_xy[0];
+  aux_local_position.position[1] = position_xy[1];
```

Comment:
Do we need some kind of check that both the `x` and `y` components have values? This goes for velocity, variances etc.
@bresch I would assume we want to avoid sending in partial positions, e.g. `position = [12.3, NAN]`.

Reply:
That's a good point, I'll add a check for this.